src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs

```diff
@@ -497,6 +499,22 @@ public static int Count<T>(this ImmutableArray<T> items, Func<T, bool> predicate
             return count;
         }
 
+        private static class UnderlyingArrayHelper<T>
+        {
+            public static readonly Func<object, object> GetValue =
+                typeof(ImmutableArray<T>).GetField("array",
+                    BindingFlags.Instance | BindingFlags.NonPublic).GetValue;
+        }
+
+        internal static T[] DangerousGetUnderlyingArray<T>(this ImmutableArray<T> array)
```

Comment:
❔ Why does the approach in **ImmutableArrayInterop.cs** not work here?

Reply:
You can't use explicit layout on generic types.

Reply:
I'd prefer we make this `private`. If there is a reason we need this to be `internal` we can re-consider. From the start though I'd prefer we expose the `ReadOnlySpan<T>` and `ReadOnlyMemory<T>` only and focus on those.

Reply:
There are two types which actually pin the underlying array using GCHandle.Alloc. I'll dig into what they're used for.

Reply:
I missed this was an existing API. If we can remove great but if not that's okay.

Reply:
Yeah, it turns out to be a huge pain for the last few bits.